service/main.go

```diff
@@ -0,0 +1,172 @@
+		return nil, err
+	}
+
+	// Open the file.
+	jsonFile, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer jsonFile.Close()
+
+	// Set defaults.
+	cfg := supervisor.Cfg{InstancesDir: "/etc/tarantool/supervisor/instances",
+		TermTimeout: 30}
+
+	// Read and parse config.
+	byteValue, err := ioutil.ReadAll(jsonFile)
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(byteValue, &cfg); err != nil {
+		return nil, err
+	}
+
+	// In the config, the time is indicated in seconds. Convert the value.
+	cfg.TermTimeout = cfg.TermTimeout * time.Second
+
+	return &cfg, nil
+}
+
+// handleSigChld handles the SIGCHLD signal.
+func handleSigChld(sv *supervisor.Supervisor) {
+	// Get PID of the terminated Instance.
+	pid, err := syscall.Wait4(-1, nil, syscall.WNOHANG, nil)
+	if err != nil || pid == 0 {
+		// If the error "no child processes" (syscall.Errno(10)) or pid == 0 -
+		// this means that the process was stopped as planned, and the wait ()
+		// function has already been called.
+		if !errors.Is(err, syscall.Errno(10)) && pid != 0 {
+			log.Printf("Can't get the PID of a process by using \"Wait4\". Error: \"%v\"",
+				err)
+		}
+		return
+	}
+	log.Printf("pid: %v\terr: %v", pid, err)
+
+	// If the isRestart Instance flag is true, try to restart the Instance.
+	if id, err := sv.RestartAfterTermInstance(pid); err != nil {
+		log.Printf("Can't restart the Instance. Old PID : %v. ID: %v. Error: \"%v\"",
+			pid, id, err)
+	} else {
+		log.Printf("The Instance has been restarted. ID: %v", id)
+	}
+}
+
+// gracefulTermination terminates the service correctly.
+func gracefulTermination(sv *supervisor.Supervisor, srv *http.Server,
+	timeout time.Duration, done chan bool) {
+	// First of all, shut down the HTTP server to avoid reciving a new request.
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	if err := srv.Shutdown(ctx); err != nil {
+		log.Printf("HTTP server shutdown error: \"%v\"", err)
+	}
+	cancel()
+
+	// And now stop all running Instances.
+	sv.StopAllInstance()
+
+	done <- true
+}
+
+// startSignalHandling adds "SIGTERM" and "SIGINT" signal handling to
+// terminate gracefully and "SIGCHLD" to restart Instances.
+func startSignalHandling(sv *supervisor.Supervisor, srv *http.Server,
+	termTimeout time.Duration, done chan bool) {
+	// The handling of the "SIGTERM" and "SIGINT" signals
+	// will be used to terminate gracefully.
+	sigTerm := make(chan os.Signal, 1)
+	signal.Notify(sigTerm, syscall.SIGINT, syscall.SIGTERM)
+
+	// The handling of the "SIGCHLD" will be used to restart Instances.
+	sigChld := make(chan os.Signal, 1)
+	signal.Notify(sigChld, syscall.SIGCHLD)
+
+	// Start signals handling.
+	go func() {
+		for {
+			select {
+			case _ = <-sigTerm:
+				gracefulTermination(sv, srv, termTimeout, done)
+			case _ = <-sigChld:
+				handleSigChld(sv)
+			}
+		}
+	}()
+}
+
+func main() {
+	// Get config.
+	args := parseArgs()
+	cfg, err := parseCfg(args.CfgPath)
```

Comment:
> Seems that parsing `cfg` is a supervisor responsibility. It also applies default config values that are supervisor-specific, but I think that `main` should only parse args, and use them to create new supervisor service.

Yeah. In the first iteration, the parsing configuration was inside the supervisor package, but after that, I thought the configuration would be more complex in the near future (and the supervisor configuration would only be a part of it), and it would be more convenient to parse it in main. I know about YAGNI, but I think I'll need it soon.

> Moreover, if `CfgPath` isn't specified we can simply use default values that are used if config file is empty.

IMHO, the presence of a config is a prerequisite for starting the service.

I think these questions are best discussed in conversation.